### DISCUSSION TITLE
Validation rules won't work

### DISCUSSION THREAD

**Discussion by robert-stanciu** [spatie/laravel-data#357] [Q&A] [answered]
Hi, community,

I am using this package validation rule inside a private package that has been generated from spatie/package-skeleton-laravel.

this is the  Data class:

```
use Spatie\LaravelData\Attributes\Validation\Rule;
use Spatie\LaravelData\Data;

class Individual extends Data
{
    public function __construct(
        #[Rule(['required', 'alpha', 'max:2'])]
        public ?string $firstName,
        public string $lastName,
        public string $socialNumber,
        public string $address,
        public string $phone,
        public string $email,
    ) {
    }
}
```

and this is how I instantiate it:
```
$individual = Individual::from([
        'firstName' => 'Long name that should fail',
        'lastName' => '',
        'socialNumber' => '',
        'address' => '',
        'phone' => '',
        'email' => '',
    ]);
```

The way I test the code is by running `composer test` (Using Pest).

There is no exception/validation done whatsoever. Is there something I am doing wrong?

**Comment by rubenvanassche** [comment, marked as answer]
Use `validateAndCreate` since `from` will only validate when a request has been passed to it.

**Reply by raveren** [reply]
this is so confusing, guys, `from` should do the validation!

**Reply by rubenvanassche** [reply]
I do not agree with this, in cases like this where you receive user input via an Http Request this would be indeed a logical default. But data is so much more and we don't need validation in a lot of cases because it adds a lot of overhead.

**Reply by raveren** [reply]
Ok so when creating a DTO class property -- I use one annotation to define name mapping, the next one to define validation, the third one to define data caster. 

Then finally when I create the DTO, one of these three identical looking annotations just doesn't work! And which one? The one which is hardest to see without explicitly checking for!

My team just now spent weeks creating a huge functionality (multiple shipping provider handling) before they realized their dtos were never validated and after partially redoing the thing (since ValidationException was never thrown and never accounted for) it turned out most of the work done relied on wrong assumptions, and now that almost all of the DTO chains are breaking due to being invalid we've been put back to square two :/ 

HUGE bummer!

Is the single argument for not validating DTOs by default - performance? Really? Performance? I'm just surprised and really sincerely can come up with **no other** arguments for it.

In my opinion this behavior is super surprising, inverse of what you would expect. 

I really think you should consider changing `from(mixed ...$payloads)` to `from(mixed $payload, bool $validate = true)` instead of the multiple payloads which are **not documented** and I have trouble using my small brain to what would you need multiple payloads for, or what would be the expected result..?? Is it for creating dtos like:

```php
SongData::from(title: 'Never gonna give you up', artist: 'Rick Astley');
```

This was possible for `spatie/data-transfer-object` and was admittedly a nice surprise that that was working, but super unintuitive and no IDE completion...

P.S. validation on instantiation was also default for `spatie/data-transfer-object`!

**Reply by raveren** [reply]
... or in fact I'd keep the `from()` signature, it has upsides, but always validate and introduce `createWithoutValidating()`

**Reply by rubenvanassche** [reply]
We have our own use cases, you have yours. This package is mainly created for our own usage and we open it to everyone to use. Don’t like some of the design decisions we made? No problem! Fork it and adapt it to your particular needs or preferences!

**Reply by raveren** [reply]
It's not about need or preference, it's what makes sense. Especially with the sub-par documentation, one would expect it to be more self-explanatory.

**Reply by raimidev** [reply]
If you want the DTO to always run validation, prob the best approach would be to create a BaseDTO in your app that extends from LaravelData and then you can just basically reuse the pipeline that Spatie has with the difference of instead of using the class-string for the `ValidatePropertiesDataPipe` you just instantiate that pipe and pass `true` as an argument in the constructor. 

```php
public static function pipeline(): DataPipeline
    {
        return DataPipeline::create()
            ->into(static::class)
            ->through(AuthorizedDataPipe::class)
            ->through(MapPropertiesDataPipe::class)
            ->through(FillRouteParameterPropertiesDataPipe::class)
            //Custom pipe to fill from bound model
            ->through(FillPropertiesFromBoundModelPipe::class)
            ->through(TrimTrailingWhitespacesFromValues::class)

           // Right here :)
            ->through(new ValidatePropertiesDataPipe(true))

            ->through(DefaultValuesDataPipe::class)
            ->through(CastPropertiesDataPipe::class);
    }
```
Our team is fine with always validating the Data since we are working in a payroll system. Hope this helps.
We have also added some custom pipes to it, so some of them might look weird.

**Reply by raveren** [reply]
It's not a very sustainable approach as you are replicating the internals of the 3rd party library in your code and when the lib changes you will have a high cost of keeping your approach up to date, or more likely, will get tied to a particular old version of this lib in the long run.

**Reply by rubenvanassche** [reply]
We'll actually the pipeline is considered something a user can do with what they want, so you're not changing internals. Could be a valid option actually.

**Comment by slnw** [comment]
needs to be fixed in my opinion.

**Reply by rubenvanassche** [reply]
See the comment from @raimirOCMI above, v4 will make it possible to always validate when using from. But the default behavior will still be disabled.
